### PR TITLE
[#4022] improvement(core): Version manange for H2 database

### PR DESCRIPTION
**Pull request by lw-yang**
### What changes were proposed in this pull request?

Version manange for H2 database like MySQL

### Why are the changes needed?

(Please clarify why the changes are needed. For instance,

Fix: #4022 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

ut

### CODE REVIEW

**Comment by lw-yang** [comment]
@yuqi1129 please help to review it

**Comment by jerryshao** [comment]
I'm curious how do you handle table schema upgrade using H2, for example if we add some new columns?

**Comment by lw-yang** [comment]
@jerryshao We use procedure to initialize data in MySQL because there is already some existing metalakes in MySQL.

However, h2 only started supporting in version 0.6.0, there is no existing metalake in the database, so there is no need for upgrade and initialization.

when upgrading from 0.6.0 to 0.7.0，we need to handle the upgrade scenario in h2

WDYT?

**Comment by jerryshao** [comment]
I think we should have a mechanism to handle this in code, not just defer this to 0.7, that's why I bring this issue out.

 The difference compared to MySQL is that for MySQL, we do the table creation/upgrading manually out of Gravitino, so we can manually handle DB upgrading and restart the Gravitino server. But for H2, because we created tables automatically in code, so there should be an automatic DB upgrading support mechanism for H2, otherwise, once we add a new table/column, how do we in-place upgrade the Gravitino.

**Comment by lw-yang** [comment]
got it

**Comment by lw-yang** [comment]
@yuqi1129 Could you please help to check if adding a README like this is okay? 

refer to Apache Hive's handling of schema upgrades for derby
https://github.com/apache/hive/blob/branch-3/metastore/scripts/upgrade/derby/README

**Comment by yuqi1129** [comment]
@lw-yang 
The following point also needs to be included in the README
1. After 0.6.0, any changes to the H2 schema must include a DDL change SQL in the script folder with a detailed issue ID in the SQL comment. 

Filename: 0.6.0-to-0.7.0-h2.sql 
content
```
# issue xxx1
ddl here


# issue xxx2 
ddl here
```

besides, can you please rename the file `README` to `Gravitino_upgrade_h2` or something similar to make the word `upgrade` visible.

**Comment by lw-yang** [comment]
> @lw-yang The following point also needs to be included in the README
> 
> 1. After 0.6.0, any changes to the H2 schema must include a DDL change SQL in the script folder with a detailed issue ID in the SQL comment.
> 
> Filename: 0.6.0-to-0.7.0-h2.sql content
> 
> ```
> # issue xxx1
> ddl here
> 
> 
> # issue xxx2 
> ddl here
> ```
> 
> besides, can you please rename the file `README` to `Gravitino_upgrade_h2` or something similar to make the word `upgrade` visible.

This README is intended for users, but `the change SQL with a detailed issue ID` should be a requirement for developers ?.  I think it should not be placed in the README.

File: scripts/h2/README [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   with the upgrade process, but others need to be resolved manually
+   or the upgrade scripts will fail to complete.
+
+5) You are now ready to run the schema upgrade scripts. If you are
+   upgrading from Gravitino 0.5.0 to Gravitino 0.6.0 you need to run the
+   upgrade-0.5.0-to-0.6.0.h2.sql script, but if you are upgrading
+   from 0.5.0 to 0.7.0 you will need to run the 0.5.0 to 0.6.0 upgrade
+   script followed by the 0.6.0 to 0.7.0 upgrade script.
+
+   % java -cp h2-1.4.200.jar org.h2.tools.RunScript -url "jdbc:h2:file:<db_file>;DB_CLOSE_DELAY=-1;MODE=MYSQL" -user <user> -password <password> -script upgrade-0.5.0-to-0.6.0-h2.sql
```
[the remark this answers is not in the document]
Comment (lw-yang) [reply scripts/h2/README:51]: the `Shell` is interactive command line tool to access a database using JDBC, we can use it

`java -cp distribution/package/libs/h2-1.4.200.jar  org.h2.tools.Shell -url jdbc:h2:file:./data/jdbc -user gravitino -password gravitino`

File: scripts/h2/README [hunk 2]
```diff
@@ -0,0 +1,64 @@
+Gravitino Upgrade HowTo
+============================
+
+This document describes how to upgrade the schema of a H2 backed
+Gravitino instance from one release version of Gravitino to another
+release version of Gravitino. For example, by following the steps listed
+below it is possible to upgrade a Gravitino 0.5.0 schema to a
```
[the remark this answers is not in the document]
Comment (lw-yang) [reply scripts/h2/README]: done

**Comment by yuqi1129** [comment]
> > @lw-yang The following point also needs to be included in the README
> > 
> > 1. After 0.6.0, any changes to the H2 schema must include a DDL change SQL in the script folder with a detailed issue ID in the SQL comment.
> > 
> > Filename: 0.6.0-to-0.7.0-h2.sql content
> > ```
> > # issue xxx1
> > ddl here
> > 
> > 
> > # issue xxx2 
> > ddl here
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > besides, can you please rename the file `README` to `Gravitino_upgrade_h2` or something similar to make the word `upgrade` visible.
> 
> This README is intended for users, but `the change SQL with a detailed issue ID` should be a requirement for developers ?. I think it should not be included in the README.

I see, let me think about it.

**Comment by yuqi1129** [comment]
@lw-yang 
Can you resolve the conflict when you have free time?

**Review by yuqi1129 — APPROVED** [review]
LGTM

File: scripts/h2/README [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+   do encounter errors you need to analyze the cause and attempt to
+   trace it back to one of the preceding steps.
+
+6) The final step of the upgrade process is validating your freshly
+   upgraded schema against the official schema for your particular
+   version of Gravitino. This is accomplished by repeating steps (3) and
+   (4), but this time comparing against the official version of the
+   upgraded schema, e.g. if you upgraded the schema to Gravitino 0.8.0 then
+   you will want to compare your schema dump against the contents of
+   schema-0.8.0-h2.sql
```
Comment (jerryshao) [inline scripts/h2/README:64]: I think it is better to have a doc about upgrading, not a README here.

**Comment by jerryshao** [comment]
@lw-yang are you still working on this?

**Comment by lw-yang** [comment]
> @lw-yang are you still working on this?

yes

**Comment by lw-yang** [comment]
@jerryshao @yuqi1129 could you please help review this again when you free

**Comment by jerryshao** [comment]
@lw-yang can you please rebase your code, we have changed the package name to `org.apache.gravitino.xxx`.

File: docs/how-to-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+### Step 5: Apply the upgrade scripts
+You are now ready to run the schema upgrade scripts. If you are
+upgrading from Gravitino 0.6.0 to Gravitino 0.7.0 you need to run the
+`upgrade-0.6.0-to-0.7.0-<type>.sql` script, but if you are upgrading
+from 0.6.0 to 0.8.0 you will need to run the 0.6.0 to 0.7.0 upgrade
+script followed by the 0.7.0 to 0.8.0 upgrade script.
+
+#### MySQL
+You can use the following command to apply the upgrade script:
```
Comment (yuqi1129) [inline docs/how-to-upgrade.md:83]: Please include information like "Assuming you are upgrading the version of Gravitino server from 0.6.0 to 0.8.0".

File: docs/index.md [hunk 5]
```diff
@@ -149,3 +149,5 @@ Gravitino provides several ways to configure and manage the Gravitino server. Se
   a Gravitino release.
 * [Publish Docker images](./publish-docker-images.md): a guide to publishing Gravitino Docker images;
   also lists the change logs of Gravitino CI Docker images and release images.
+* [How to upgrade Gravitino](./how-to-upgrade.md): a guide to upgrading the schema of backed
+  Gravitino instance from one release version to another.
```
Comment (yuqi1129) [inline docs/index.md:153]: of Gravitino storage backend

**Review by yuqi1129 — COMMENTED** [review]
Just these minor issues, others LGTM

**Comment by lw-yang** [comment]
done

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Comment by yuqi1129** [comment]
@lw-yang 
Could you please resolve the conflicts?

**Comment by lw-yang** [comment]
done

**Pull request merged**